Framework/API/test/WorkspacePropertyWithIndexTest.h

```diff
@@ -0,0 +1,319 @@
+#ifndef MANTID_API_WORKSPACEPROPERTYWITHINDEXTEST_H_
+#define MANTID_API_WORKSPACEPROPERTYWITHINDEXTEST_H_
+
+#include "MantidAPI/WorkspaceFactory.h"
+#include "MantidTestHelpers/FakeObjects.h"
+#include <boost/shared_ptr.hpp>
+#include <cxxtest/TestSuite.h>
+
+#include "MantidAPI/WorkspaceProperty.tcc"
+#include "MantidAPI/WorkspacePropertyWithIndex.h"
+#include "MantidAPI/WorkspacePropertyWithIndex.tcc"
+#include "MantidKernel/PropertyManager.h"
+#include "MantidKernel/make_unique.h"
+#include <numeric>
+
+using namespace Mantid::API;
+using Mantid::API::WorkspacePropertyWithIndex;
+
+namespace {
+class PropertyManagerHelper : public PropertyManager {
+public:
+  PropertyManagerHelper() : PropertyManager() {}
+
+  using PropertyManager::declareProperty;
+  using PropertyManager::setProperty;
+  using PropertyManager::getPointerToProperty;
+};
+} // namespace
+
+using MatrixWorkspaceIndexProp = WorkspacePropertyWithIndex<MatrixWorkspace>;
+class WorkspacePropertyWithIndexTest : public CxxTest::TestSuite {
+public:
+  // This pair of boilerplate methods prevent the suite being created statically
+  // This means the constructor isn't called when running other tests
+  static WorkspacePropertyWithIndexTest *createSuite() {
+    return new WorkspacePropertyWithIndexTest();
+  }
+  static void destroySuite(WorkspacePropertyWithIndexTest *suite) {
+    delete suite;
+  }
+
+  void testConstructor() {
+    TS_ASSERT_THROWS_NOTHING((MatrixWorkspaceIndexProp()));
+  }
+
+  void testConstructorFailsWithInvalidIndexType() {
+    TS_ASSERT_THROWS(MatrixWorkspaceIndexProp("InputWorkspaceWithIndex", 0),
+                     std::invalid_argument);
```

Comment:
Why do you need to cast in the client code, instead of within `WorkspacePropertyWithIndex`, before passing the values on to the internal `ArrayProperty`?